Fix memory leak in AsyncCompletions.parse() with dynamically created models

mousberg opened the pull request:
Fixes #2146

This PR fixes a memory leak in the manually maintained library code (src/openai/lib/_parsing/_completions.py) where AsyncCompletions.parse() was retaining schema representations of dynamically created Pydantic models indefinitely.

### Changes Made:
- Implemented WeakKeyDictionary cache for schema objects in _parsing/_completions.py
- Added comprehensive test suite in tests/lib/_parsing/test_memory_leak.py

### Technical Details:
The fix uses Python's WeakKeyDictionary to store schema representations, allowing them to be garbage collected when their corresponding model types are no longer referenced. This prevents the unbounded memory growth observed when repeatedly calling parse() with new models created via create_model().

### Test Coverage:
New test suite verifies:
- Schema cache properly handles dynamic models
- Memory is released when models are no longer referenced
- Both synchronous and asynchronous usage patterns

### References:
- Issue: #2146
- Memory leak reproduction case provided in the issue

anteverse commented on src/openai/lib/_parsing/_completions.py:
```diff
@@ -28,6 +29,9 @@
 from ...types.chat.completion_create_params import ResponseFormat as ResponseFormatParam
 from ...types.chat.chat_completion_message_tool_call import Function
 
+# Cache to store weak references to schema objects
+_schema_cache = weakref.WeakKeyDictionary()
```
Hi @mousberg ,

Thanks for digging into this. As I understand this cache could help in the optimization of `type_to_response_format_param` as it will not recompute an already existing schema. I guess it solves the issue [2146(]https://github.com/openai/openai-python/issues/2146) for the given example. But I don't see how it could stop the memory from rising up when we submit models that are always different.
Could we use some kind of size max size for this cache? Something like an LRU? 
Your first test case actually brings a very good insight: if we keep on generating new models, the cache size grows infinitely.

anteverse replied:
Hey,
Thanks for narrowing down the problems, I think you did most of the job.

> digging into an LRU-based approach, I'd love to collaborate!

I'd love to collaborate as well. 

> ran into challenges with picking the right cache key strategy (model identity vs. schema-based hashing)

By any chance, do you have any connections at Pydantic? The folks there could help come up with a common key.

By the way, if we end up with a cache that has a limited size, what would be a good measure for the limit? Should it be frozen? Accessible to the end user? Should we have an heuristic based on the machine's capacity?

I'll run a few tests on my end as well!

mousberg replied:
@anteverse Yeah, reached out to Samuel (Pydantic's founder), and he suggested an approach for the cache key:
Build a string from the fields and types, optionally hash it, and use that as the cache key.

He also mentioned that if we need more input, it might be worth posting on Pydantic's GitHub to see if the community has additional insights.

Thoughts? I can prototype this approach and see if it addresses the issue!

RobertCraigie replied:
> But I don't see how it could stop the memory from rising up when we submit models that are always different.
Could we use some kind of size max size for this cache?

maybe I'm missing something but I don't see how this is a concern? if you're continuously creating new models then they should be getting GC'd at some point and if they're not, then you're still holding a reference to the model type yourself? if that is the case can you share more about your use case?

RobertCraigie commented on tests/lib/_parsing/test_memory_leak.py:
```diff
@@ -0,0 +1,101 @@
+    _schema_cache.clear()
+    initial_cache_size = len(_schema_cache)
+    
+    # Create a mock client 
+    mock_client = MagicMock()
+    mock_client.chat.completions.create = AsyncMock()
+    
+    # Create the AsyncCompletions instance with our mock client
+    completions = AsyncCompletions(mock_client)
```
this seems to be unused?

RobertCraigie commented on tests/lib/_parsing/test_memory_leak.py:
```diff
@@ -0,0 +1,101 @@
+from openai.lib._parsing import type_to_response_format_param
+from openai.lib._parsing._completions import _schema_cache
+
+class TestMemoryLeak(unittest.TestCase):
```
we do not use unittest, can you delete this? it seems to be the same as the pytest version anyway

RobertCraigie commented on src/openai/lib/_parsing/_completions.py:
```diff
@@ -254,11 +262,16 @@ def type_to_response_format_param(
         raise TypeError(f"Unsupported response_format type - {response_format}")
 
-    return {
+    schema_param = {
```
nit: 
```suggestion
    schema_param: ResponseFormatParam = {
```

RobertCraigie reviewed (changes requested):
This approach seems good but the tests need cleaning up.

mousberg commented:
@RobertCraigie Thanks for the feedback! I've addressed all the requested changes:

- Removed the unittest class and switched to using pytest exclusively
- Removed unused mock client and AsyncCompletions instance
- Added proper type annotation for schema_param: `schema_param: ResponseFormatParam = {`
- Fixed all lint issues in the test file

The tests are now passing and the lint checks are clean. Let me know if I've missed anything

mousberg commented:
Thanks @karpetrosyan for identifying the runtime generic types as the root cause. Updated this PR to fix all instances: `_completions.py`, `_responses.py`, and the streaming `ContentDoneEvent` in `streaming/chat/_completions.py`.

karpetrosyan commented:
> Wait so this guy [David Gelberg](https://github.com/mousberg) spend time writing this and you guys just ignore it and copy the code in another PR so a original contributor of the repo can farm a PR merge? Open source is wild

Pay attention to the commit dates, please. 
The original changes in this PR were completely different, so we re-implemented them in another PR to avoid changing everything in this one. We appreciate the time and effort of each contributor.
